src/index.ml

```diff
@@ -457,34 +490,43 @@ module Make (K : Key) (V : Value) (IO : IO) = struct
     let t = assert_open t in
     Log.debug (fun l -> l "add %a %a" K.pp key V.pp value);
     if t.config.readonly then raise RO_not_allowed;
-    append_key_value t.log key value;
-    Tbl.replace t.log_mem key value;
-    if Int64.compare (IO.offset t.log) (Int64.of_int t.config.log_size) > 0
+    let log = assert_and_get t.log in
+    append_key_value log.io key value;
+    Tbl.replace log.mem key value;
+    if Int64.compare (IO.offset log.io) (Int64.of_int t.config.log_size) > 0
     then merge ~witness:{ key; key_hash = K.hash key; value } t
 
   let iter f t =
     let t = assert_open t in
     Log.debug (fun l -> l "iter %S" t.root);
     if t.config.readonly then sync_log t;
-    Tbl.iter f t.log_mem;
-    may (fun index -> iter_io (fun e -> f e.key e.value) index.io) t.index
+    match t.log with
+    | None -> ()
+    | Some log ->
+        Tbl.iter f log.mem;
+        may
+          (fun (i : index) -> iter_io (fun e -> f e.key e.value) i.io)
+          t.index
+
+  let flush_instance instance =
+    if instance.config.readonly then raise RO_not_allowed;
+    let log = assert_and_get instance.log in
+    IO.sync log.io
 
   let flush t =
-    let t = assert_open t in
-    IO.sync t.log
+    let instance = assert_open t in
+    flush_instance instance
 
   let close it =
     match !it with
     | None -> ()
     | Some t ->
+        it := None;
         t.open_instances <- t.open_instances - 1;
```

Comment:
As discussed with @CraigFe and @icristescu, there is a concurrency issue here. The pattern-matching + set to `None` should be atomic.
The easiest way would be to `Mutex.lock`, that adds a `threads` dependency but it will probably be added anyway with the concurrent merges until we properly functorize over the IO.
Another way is to just merge as-is since thread safety was already killed by https://github.com/mirage/index/pull/108.

Reply:
Don't worry about concurrency issues right now -- a comment in the code is enough.